src/ScalarMessaging.ts

```diff
@@ -337,6 +340,15 @@ function setWidget(event: MessageEvent<any>, roomId: string): void {
             sendError(event, _t("Unable to create widget."), new Error("Optional field 'data' must be an Object."));
             return;
         }
+        if (widgetAvatarUrl !== undefined && typeof widgetAvatarUrl !== 'string'
+            && !(widgetAvatarUrl instanceof Blob)) {
+            sendError(
+                event,
+                _t("Unable to create widget."),
+                new Error("Optional field 'avatar_url' must be a string or Blob."),
+            );
+            return;
```

Comment:
The spec seems to say it must be an mxc string?

> A new optional paramater named `avatar_url` is added to the widget definition. This parameter is
an MXC URI to an image clients can use to associate with the widget, likely alongside the `name`
and/or `title`.

Reply:
Yeah, the event is specified to use an mxc: URI. But this is about this API that an integration manager uses. Just having the ability to set an mxc: URI would not be sufficient as there is no way for an integration manager to upload a file to the room. As an alternative, we could just pass a string `avatar_url` here and add an additional `upload_file` request to the API. Integration managers can then chain the two calls to get to the same result.

Reply:
I don't think that there is a spec for the whole API at all, at least I couldn't find any. The only available "spec" or better documentation is at the beginning of this file and in Dimension: https://github.com/turt2live/matrix-dimension/blob/master/docs/reference/scalar_client_api.md

Reply:
The postMessage API is capable of passing a Blob, though for API semantics we should be requiring a MSC URI over this transport. The widget API (or in this case, the integration manager API) doesn't currently have much ability to make use of a user's theoretical quota for media uploads, and I don't think we should let it here.

Instead, the integration manager would be expected to already know the MXC URI it wants to use (either because it has a bot account to upload the media or because it was hardcoded in a config somewhere) - the upload isn't tied to a room, so can be safely reused. This also helps with homeserver administration because the homeserver admin can flag the upload as "safe" and never delete it, even when purging a whole room's worth of content. It also means homeservers aren't storing duplicate copies of the content.

Reply:
Good points - as long there is no way to detect whether the avatar was already uploaded we would upload it again and again…

Then let's start with only an mxc: URI 👍 .